docs/shader-modules/project.md

```diff
@@ -18,94 +20,120 @@ attribute float instanceSize;
+  gl_Position = project_common_position_to_clipspace(center + vertex);
 }
 ```
 
 ## getUniforms
 
-The JavaScript uniforms are extracted mainly from the viewport with a few additional parameters (which deck.gl supplies from `LayerManager` state or `Layer` props of course).
-
-Provided by `LayerManager`:
+The JavaScript uniforms are extracted from both contextual and per-layer props. The following information are used:
 
 * `viewport`
 * `devicePixelRatio`
-
-Provided by `Layer` props:
-
 * `coordinateSystem`
 * `coordinateOrigin`
 * `modelMatrix`
 
 
 ## GLSL Uniforms
 
-Uniform names are not considered part of the official API of shader modules and can potentially change between minor releases, but are documented for applications that need this level of access. Use the module's public GLSL functions instead of directly accessing uniforms when possible.
-
-| Uniform | Type | Description |
-| --- | --- | --- |
-| project_uModelMatrix | mat4 | model matrix (identity if not supplied) |
-| project_uViewProjectionMatrix | mat4 | combined view projection matrix |
-| project_uViewportSize | vec2 | size of viewport in pixels |
-| project_uDevicePixelRatio | float | device pixel ratio of current viewport (value depends on `useDevicePixels` prop) |
-| project_uFocalDistance | float | distance where "pixel sizes" are display in 1:1 ratio (modulo `devicePixelRatio`) |
-| project_uCameraPosition | vec3 | position of camera in world space |
-| project_uCoordinateSystem | float | COORDINATE_SYSTEM enum |
-| project_uCenter | float | coordinate origin in world space |
-| project_uScale | float | Web Mercator scale (2^zoom) |
-| project_uPixelsPerMeter | vec3 | Web Mercator pixels per meter near the current viewport center |
-| project_uPixelsPerDegree | vec3 | Web Mercator pixels per degree near the current viewport center |
+Uniforms are considered private to each shader module. They may change in between patch releases. Always use documented functions instead of accessing module uniforms directly.
 
+The uniforms of the `project` shader module are prefixed with `project_` in their names.
 
 ## GLSL Functions
 
 The projection module makes it easy to write vertex shaders that follow deck.gl's projection methods, enabling your layer to accept coordinates in both [longitude,latitude,altitude] or [metersX,metersY,metersZ] format. To support the basic features expected of a deck.gl layer, such as various viewport types and coordinate systems, your own shaders should always use the built-in projection functions.
 
+The functions converts positions/vectors between 4 coordinate spaces:
+
+* World space (`world`): the [coordinate system](/docs/developer-guide/coordinate-systems.md) defined by the layer, not necessarily linear or uniform.
+* Common space (`common`): an normalized intermediate 3D space that deck.gl uses for consistent processing of geometries. It is safe to add/subtract/scale positions and vectors in this space.
```

Comment:
`an normalized intermediate 3D space1`

* Typo `an` -> `a`
* Nit: Maybe add the word `linear` here?